File: guides/micronaut-aws-lambda-s3-event/java/src/main/java/example/micronaut/FunctionRequestHandler.java
package example.micronaut;

import com.amazonaws.services.lambda.runtime.events.models.s3.S3EventNotification;
import io.micronaut.serde.annotation.Serdeable;
import io.micronaut.function.aws.MicronautRequestHandler;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import software.amazon.awssdk.core.sync.RequestBody;
import software.amazon.awssdk.services.s3.S3Client;
import software.amazon.awssdk.services.s3.model.GetObjectRequest;
import software.amazon.awssdk.services.s3.model.PutObjectRequest;
import jakarta.inject.Inject;
import java.util.Locale;

@Serdeable
public class FunctionRequestHandler
        extends MicronautRequestHandler<S3EventNotification, Void> { // <1>
    private static final Logger LOG = LoggerFactory.getLogger(FunctionRequestHandler.class);
    private static final String SLASH = "/";
    private static final String THUMBNAILS = "thumbnails";
    public static final String OBJECT_CREATED = "ObjectCreated";
    private static final String JPG = "jpg";
    private static final String PNG = "png";
    private static final char DOT = '.';

    @Inject
    public S3Client s3Client; // <2>

    @Inject
    public ThumbnailGenerator thumbnailGenerator; // <3>

    @Override
    public Void execute(S3EventNotification input) {
        for (S3EventNotification.S3EventNotificationRecord record : input.getRecords()) {

            LOG.info("event name: {}" , record.getEventName());
            if (record.getEventName().contains(OBJECT_CREATED)) { // <4>
                S3EventNotification.S3Entity s3Entity = record.getS3();
                String bucket = s3Entity.getBucket().getName();
                String key = s3Entity.getObject().getKey();
                int index = key.lastIndexOf(DOT);
                if (index != -1) {
                    String format = key.substring(index + 1).toLowerCase(Locale.ENGLISH);
                    if (format.equals(PNG) || format.equals(JPG)) {
                        thumbnailGenerator.thumbnail(s3Client.getObject(GetObjectRequest.builder()
                                .bucket(bucket)
                                .key(key)
                                .build()), format)
                                .ifPresent(bytes -> s3Client.putObject(PutObjectRequest.builder()
                                        .key(thumbnailKey(key))
                                        .bucket(bucket)
                                        .build(), RequestBody.fromBytes(bytes)));
                    }
                }
            }
        }
        return null;
    }

    private static String thumbnailKey(String key) {
        int index = key.lastIndexOf(SLASH);
        String fileName = index != -1 ? key.substring(index + SLASH.length()) : key;
        return THUMBNAILS + SLASH + fileName;
    }
}

File: guides/micronaut-aws-lambda-s3-event/kotlin/src/main/kotlin/example/micronaut/FunctionRequestHandler.kt
/*
 * Copyright 2017-2026 original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package example.micronaut

import com.amazonaws.services.lambda.runtime.events.models.s3.S3EventNotification
import io.micronaut.function.aws.MicronautRequestHandler
import io.micronaut.serde.annotation.Serdeable
import jakarta.inject.Inject
import org.slf4j.LoggerFactory
import software.amazon.awssdk.core.sync.RequestBody
import software.amazon.awssdk.services.s3.S3Client
import software.amazon.awssdk.services.s3.model.GetObjectRequest
import software.amazon.awssdk.services.s3.model.PutObjectRequest
import java.util.Locale

@Serdeable
class FunctionRequestHandler : MicronautRequestHandler<S3EventNotification, Void?>() { // <1>

    @Inject
    lateinit var s3Client: S3Client // <2>

    @Inject
    lateinit var thumbnailGenerator: ThumbnailGenerator // <3>

    override fun execute(input: S3EventNotification): Void? {
        for (record in input.records) {
            LOG.info("event name: {}", record.eventName)
            if (record.eventName.contains(OBJECT_CREATED)) { // <4>
                val s3Entity = record.s3
                val bucket = s3Entity.bucket.name
                val key = s3Entity.`object`.key
                val index = key.lastIndexOf(DOT)
                if (index != -1) {
                    val format = key.substring(index + 1).lowercase(Locale.ENGLISH)
                    if (format == PNG || format == JPG) {
                        thumbnailGenerator.thumbnail(
                            s3Client.getObject(
                                GetObjectRequest.builder()
                                    .bucket(bucket)
                                    .key(key)
                                    .build()
                            ),
                            format
                        )?.let { bytes ->
                            s3Client.putObject(
                                PutObjectRequest.builder()
                                    .key(thumbnailKey(key))
                                    .bucket(bucket)
                                    .build(),
                                RequestBody.fromBytes(bytes)
                            )
                        }
                    }
                }
            }
        }
        return null
    }

    private fun thumbnailKey(key: String): String {
        val index = key.lastIndexOf(SLASH)
        val fileName = if (index != -1) key.substring(index + SLASH.length) else key
        return "$THUMBNAILS$SLASH$fileName"
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(FunctionRequestHandler::class.java)
        private const val SLASH = "/"
        private const val THUMBNAILS = "thumbnails"
        const val OBJECT_CREATED = "ObjectCreated"
        private const val JPG = "jpg"
        private const val PNG = "png"
        private const val DOT = '.'
    }
}
